src/main/java/org/opensearch/flowframework/util/TenantAwareHelper.java

```diff
@@ -88,4 +94,156 @@ public static String getTenantID(Boolean isMultiTenancyEnabled, RestRequest rest
+     * @param tenantId The ID of the tenant requesting the deprovision.
+     * @param workflowListener The listener to notify in case of failure.
+     * @return true if the deprovision slot was acquired, false otherwise.
+     */
+    public static boolean tryAcquireDeprovision(int maxExecutions, String tenantId, ActionListener<WorkflowResponse> workflowListener) {
+        if (!tryAcquire(tenantId, activeDeprovisionsPerTenant, maxExecutions)) {
+            workflowListener.onFailure(
+                new FlowFrameworkException(
+                    "Exceeded max simultaneous deprovisioning requests: " + maxExecutions,
+                    RestStatus.TOO_MANY_REQUESTS
+                )
+            );
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Releases a provision slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the provision slot.
+     */
+    public static void releaseProvision(String tenantId) {
+        release(tenantId, activeProvisionsPerTenant);
+    }
+
+    /**
+     * Releases a deprovision slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the deprovision slot.
+     */
+    public static void releaseDeprovision(String tenantId) {
+        release(tenantId, activeDeprovisionsPerTenant);
+    }
+
+    /**
+     * Attempts to acquire an execution slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant requesting the execution slot.
+     * @param executionsMap The map tracking the number of active executions per tenant.
+     * @param maxExecutions The maximum number of simultaneous executions allowed per tenant.
+     * @return true if the execution slot was acquired, false otherwise.
+     */
+    private static boolean tryAcquire(String tenantId, ConcurrentHashMap<String, AtomicInteger> executionsMap, int maxExecutions) {
+        if (tenantId == null) {
+            return true; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.computeIfAbsent(tenantId, k -> new AtomicInteger(0));
+        if (count.incrementAndGet() <= maxExecutions) {
+            return true;
+        } else {
+            count.decrementAndGet(); // Rollback the increment
+            return false;
+        }
+    }
+
+    /**
+     * Releases an execution slot for the given tenant.
+     *
+     * @param tenantId The ID of the tenant for which to release the execution slot.
+     * @param executionsMap The map tracking the number of active executions per tenant.
+     */
+    private static void release(String tenantId, ConcurrentHashMap<String, AtomicInteger> executionsMap) {
+        if (tenantId == null) {
+            return; // No throttling for null tenantId
+        }
+        AtomicInteger count = executionsMap.get(tenantId);
+        if (count != null) {
+            count.decrementAndGet();
+        }
```

Comment:
Don't we have to remove tenantId from the map here?
```suggestion
        if (count != null && count.decrementAndGet() <= 0) {
        executionsMap.remove(tenantId);
    }
```

Reply:
The code logic would work the same here.  But I can see how filling up a static map with a bunch of 0's could leave a bunch of zero map entries.

Reply:
Your code suggestion still leaves a race condition between the decrement and get, and the removal.

I think we could use computeIfPresent, though, and it removes the entry on null.

Reply:
+1 on computeIfPresent so if count is 0 then we should remove entry from map

Reply:
Additionally could a race condition here cause a negative count?

Reply:
> could a race condition here cause a negative count?

I don't think so?  The code to decrement is only run after there was an increment.